test/e2e/app/app.go

```diff
@@ -533,22 +566,36 @@ func parseTx(tx []byte) (string, string, error) {
 	return string(parts[0]), string(parts[1]), nil
 }
 
-func (app *Application) verifyAndSum(currentHeight int64, extCommit *abci.ExtendedCommitInfo, callsite string) (int, int64, error) {
+func (app *Application) verifyAndSum(areExtensionsEnabled bool, currentHeight int64, extCommit *abci.ExtendedCommitInfo, callsite string) (int64, error) {
 	var sum int64
 	var extCount int
 	for _, vote := range extCommit.Votes {
-		if !vote.SignedLastBlock || len(vote.VoteExtension) == 0 {
+		if !vote.SignedLastBlock {
+			continue
+		}
+		if !areExtensionsEnabled {
+			if len(vote.VoteExtension) != 0 {
+				return 0, fmt.Errorf("non-empty vote extension at height %d, which has extensions disabled", currentHeight)
+			}
+			continue
+		}
+		if len(vote.VoteExtension) == 0 {
```

Comment:
One last question related to the comment above. Here the length is allowed to be 0 (as it is supposed to be), but then it will fail later? I am just asking to confirm that you don't want to impose your more stringent condition here as well.

Reply:
That's an excellent question. :-)
The answer is coming in one of my next PRs. The problem is that the current ABCI data structures can only say if the sender of the vote "signed or not" (i.e., a boolean). But if the sending validator has signed for `nil`, `SignedLastBlock` will also be true; however as it is a `nil` precommit it won't have an extension.
It took me a few hours of troubleshooting with e2e tests to find this out.
We discussed this in @ebuchman's office hours on Monday, and we decided we're gonna fix the ABCI interface (a boolean is not enough), since anyway we're introducing breaking changes.
This is what's coming in one of my next PRs.

Reply:
#540